### DISCUSSION TITLE
Support/replacement for startup-scripts?

### DISCUSSION THREAD

**Discussion by famod** [keycloak/keycloak#12058] [Archive]
Hi all,

I have just started looking at how to migrate from legacy KC 18 to KC.X.

Currently, we are using a shell script in `/opt/jboss/startup-scripts` to initialize our realm.
We are basically using this approach: https://keycloak.discourse.group/t/run-script-after-server-startup/5043/6

Via our script we added a simple "migration" logic that is updating/evaluating a custom real attribute holding an "init number".
This way, by wrapping realm modifications in blocks that are given increasing numbers, we are able to migrate out realm incrementally.
(I can provide more details if needed.)

Now, are such `startup-scripts` even supported in KC.X?
I found nothing in the guides, but I found this https://keycloak.discourse.group/t/keycloak-x-docker-startup-scripts-not-being-executed/8208 (which is rather old though, from the very early days of KC.X).

**Comment by famod** [comment]
FTR, another user has posted his solution here: https://keycloak.discourse.group/t/keycloak-x-docker-startup-scripts-not-being-executed/8208/6?u=famod

I came up with a very simple workaround:

In `Dockerfile`:
```
ENTRYPOINT [ "/opt/myproject/kc-myproject.sh" ]
CMD ["start"]
```
which overrides the default `ENTRYPOINT [ "/opt/keycloak/bin/kc.sh" ]`.

`kc-myproject.sh` looks like this:
```
#!/bin/bash
# Wrapper script as docker entrypoint to run initialize-my-realm.sh in parallel to actual kc.sh (the official entrypoint).

set -e -u -o pipefail
shopt -s failglob

/opt/myproject/initialize-my-realm.sh & disown

/opt/keycloak/bin/kc.sh "$@"
```

`initialize-my-realm.sh` then polls KC to check whether it's ready to called via `kcadm.sh`.

**Reply by NghiemHieu7295** [reply]
So, to specify `initialize-my-realm.sh` must be wait for `kc.sh` finish starting, how can we do?

**Comment by razholio** [comment]
you can always overload 'command:' thusly:
command:
  - "bash"
  - "-c"
  - "cp /etc/pki/ca-trust/extracted/java/cacerts /etc/x509/ && keytool -import -trustcacerts -file /etc/x509/server.crt -keystore /etc/x509/cacerts -alias server.crt -storepass changeit -noprompt && /opt/keycloak/bin/kc.sh start --http-enabled=true --http-port=8080 --spi-truststore-file-file=/etc/x509/cacerts --spi-truststore-file-password=changeit --spi-truststore-file-hostname-verification-policy=ANY --https-certificate-file=/etc/x509/custom.crt --https-certificate-key-file=/etc/x509/custom.key"
